File: nattybear/level2/튜플/Solution.java
import java.util.*;
import java.util.stream.*;

class Solution {
  public int[] solution(String s) {
    Calculator calculator = new Calculator(s);
    List<Integer> answer = new ArrayList<>();
    List<Set<Integer>> list = calculator.getSets();
    list
      .stream()
      .sorted((x, y) -> x.size() - y.size())
      .reduce(new HashSet<Integer>(), (acc, x) -> {
        Set<Integer> temp = new HashSet<>();
        temp.addAll(x);
        x.removeAll(acc);
        answer.add(x.iterator().next());
        return temp;
      });
    return answer.stream().mapToInt(x -> x).toArray();
  }
}

class Calculator {
  private String expression;
  private int current;
  private char look;

  public Calculator(String e) {
    expression = e;
    current = 0;
    getChar();
  }

  public void getChar() {
    if (current < expression.length())
      look = expression.charAt(current++);
  }

  public int getNum() {
    StringBuilder builder = new StringBuilder();
    while (Character.isDigit(look)) {
      builder.append(look);
      getChar();
    }
    return Integer.parseInt(builder.toString());
  }

  public Set<Integer> getSet() {
    Set<Integer> set = new HashSet<>();
    getChar();
    set.add(getNum());
    while (look != '}') {
      getChar();
      set.add(getNum());
    }
    getChar();
    return set;
  }

  public List<Set<Integer>> getSets() {
    List<Set<Integer>> sets = new ArrayList<>();
    getChar();
    sets.add(getSet());
    while (look != '}') {
      getChar();
      sets.add(getSet());
    }
    return sets;
  }
}

File: nattybear/level2/튜플/solution.js
class Calculator {
  constructor(expression) {
    this.expression = expression
    this.current = 0
    this.get_char()
  }

  get_char() {
    if (this.current < this.expression.length)
      this.look = this.expression.charAt(this.current++)
  }

  get_num() {
    let tmp = ''
    while (/[0-9]/.test(this.look)) {
      tmp += this.look
      this.get_char()
    }
    return parseInt(tmp)
  }

  get_set() {
    const set = new Set()
    this.get_char()
    set.add(this.get_num())
    while (this.look != '}') {
      this.get_char()
      set.add(this.get_num())
    }
    this.get_char()
    return set
  }

  get_sets() {
    const sets = []
    this.get_char()
    sets.push(this.get_set())
    while (this.look != '}') {
      this.get_char()
      sets.push(this.get_set())
    }
    return sets
  }
}

function solution(s) {
  const calculator = new Calculator(s)
  const answer = []
  const sets = calculator.get_sets()
  sets.sort((x, y) => x.size - y.size)
  sets.reduce((x, y) => {
    const tmp = new Set(y)
    const set = new Set([...y].filter(z => !x.has(z)))
    set.forEach((k, _) => answer.push(k))
    return tmp
  }, new Set())
  return answer
}
